plugins/3scale-provider/src/providers/ThreeScaleApiEntityProvider.ts

```diff
@@ -0,0 +1,229 @@
+  private connection?: EntityProviderConnection;
+
+  static fromConfig(
+    configRoot: Config,
+    options: {
+      logger: Logger;
+      schedule?: TaskRunner;
+      scheduler?: PluginTaskScheduler;
+    },
+  ): ThreeScaleApiEntityProvider[] {
+    const providerConfigs = readThreeScaleApiEntityConfigs(configRoot);
+
+    if (!options.schedule && !options.scheduler) {
+      throw new Error('Either schedule or scheduler must be provided.');
+    }
+
+    return providerConfigs.map(providerConfig => {
+      if (!options.schedule && !providerConfig.schedule) {
+        throw new Error(
+          `No schedule provided neither via code nor config for ThreeScaleApiEntityProvider:${providerConfig.id}.`,
+        );
+      }
+
+      const taskRunner =
+        options.schedule ??
+        options.scheduler!.createScheduledTaskRunner(providerConfig.schedule!);
+
+      return new ThreeScaleApiEntityProvider(
+        providerConfig,
+        options.logger,
+        taskRunner,
+      );
+    });
+  }
+
+  private constructor(
+    config: ThreeScaleConfig,
+    logger: Logger,
+    taskRunner: TaskRunner,
+  ) {
+    this.env = config.id;
+    this.baseUrl = config.baseUrl;
+    this.accessToken = config.accessToken;
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+
+    this.scheduleFn = this.createScheduleFn(taskRunner);
+  }
+
+  private createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:run`;
+      return taskRunner.run({
+        id: taskId,
+        fn: async () => {
+          try {
+            await this.run();
+          } catch (error) {
+            this.logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  getProviderName(): string {
+    return `ThreeScaleApiEntityProvider:${this.env}`;
+  }
+
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
+  }
+
+  async run(): Promise<void> {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    this.logger.info(`Discovering ApiEntities from 3scale ${this.baseUrl}`);
+
+    const entities: Entity[] = [];
+
+    let page: number = 0;
+    let services: Services;
+    let apiDocs: APIDocs;
+    let fetchServices: boolean = true;
+    while (fetchServices) {
+      services = await listServices(
+        this.baseUrl,
+        this.accessToken,
+        page,
+        ThreeScaleApiEntityProvider.SERVICES_FETCH_SIZE,
+      );
+      apiDocs = await listApiDocs(this.baseUrl, this.accessToken);
+      for (const element of services.services) {
+        const service = element;
+        this.logger.debug(`Find service ${service.service.name}`);
+
+        // Trying to find the API Doc for the service.
+        const apiDoc = apiDocs.api_docs.find(obj => {
+          return obj.api_doc.service_id === service.service.id;
+        });
```

Comment:
I changed this a little to validate if api_doc_service_id is undefined, this is because it is possible to have an Active Doc not associated with a product on 3scale.